rs/execution_environment/src/canister_manager.rs

```diff
@@ -2031,7 +2034,277 @@ impl CanisterManager {
+        }
+
+        if self.config.rate_limiting_of_heap_delta == FlagStatus::Enabled
+            && canister.scheduler_state.heap_delta_debit >= self.config.heap_delta_rate_limit
+        {
+            return Err(CanisterManagerError::WasmChunkStoreError {
+                message: format!(
+                    "Canister is heap delta rate limited. Current delta debit: {}, limit: {}",
+                    canister.scheduler_state.heap_delta_debit, self.config.heap_delta_rate_limit
+                ),
+            });
+        }
+
+        // Write data to the appropriate location, as specified by the `CanisterSnapshotDataOffset` variant.
+        // Memory has already been paid for by `create_snapshot_from_metadata`,
+        // but the instructions used to copy the data still need to be accounted for.
+        let snapshot_inner = Arc::make_mut(snapshot);
+        let num_bytes = match args.kind {
+            CanisterSnapshotDataOffset::WasmModule { offset } => {
+                let res = snapshot_inner
+                    .execution_snapshot_mut()
+                    .wasm_binary
+                    .write(&args.chunk, offset as usize);
+                if res.is_err() {
+                    return Err(CanisterManagerError::InvalidSubslice {
+                        offset,
+                        size: args.chunk.len() as u64,
+                    });
+                }
+                args.chunk.len() as u64
+            }
+            CanisterSnapshotDataOffset::MainMemory { offset } => {
+                let mut buffer = Buffer::new(snapshot_inner.wasm_memory().page_map.clone());
+                buffer.write(&args.chunk, offset as usize);
+                args.chunk.len() as u64
+            }
+            CanisterSnapshotDataOffset::StableMemory { offset } => {
+                let mut buffer = Buffer::new(snapshot_inner.stable_memory().page_map.clone());
+                buffer.write(&args.chunk, offset as usize);
+                args.chunk.len() as u64
+            }
+            CanisterSnapshotDataOffset::WasmChunk => {
+                // The chunk store is initialized as empty, and no memory cost has been charged yet.
+                // So we charge for both memory and instructions here.
+                let validated_chunk = match snapshot_inner
+                    .chunk_store_mut()
+                    .can_insert_chunk(self.config.wasm_chunk_store_max_size, args.chunk.clone())
+                {
+                    ChunkValidationResult::Insert(validated_chunk) => validated_chunk,
+                    ChunkValidationResult::AlreadyExists(hash) => {
+                        return Err(CanisterManagerError::WasmChunkStoreError {
+                            message: format!(
+                                "Chunk with hash {:02x?} already present in chunk store.",
+                                hash
+                            ),
+                        })
```

Comment:
The (somewhat similar) endpoint `upload_chunk` would return with a reply in this case (i.e., it wouldn't fail).